### ISSUE TITLE
v03 heads up.

### ISSUE THREAD

**Issue by petersilva** [TheTannerRyan/ecpush#2]
just to let you know, for the past year or so, we have been working on a new message payload format, as a result of limitations in the current ones and feed back from some international consultations. The current protocol version is identified by using the topic tree that starts with v02.post. Over the next year or two, we may migrate to v03.post. Differences in messages:

AMQP headers are no longer used to store key-value pairs. Instead, the message body is a JSON array. As a result, the formerly anonymous fields in the body of a v02 message are now key value pairs in the array: pubTime, baseUrl, and relPath.

for fields that are encoded, such as checksums, the encoding is changed to base64 (more compact representation)

https://github.com/MetPX/sarracenia/blob/master/doc/sr_postv3.7.rst

might still evolve slightly (new fields?) but we have done some important deployments of v03, and it is looking solid. no fire... nothing will be sprung on consumers suddenly, we haven´t looked at any migration strategy yet, but would not want to spring it on clients all of a sudden. Figured you would want to know far ahead of time.

I can supply some alternate data streams if you want a sample.

**Comment by tannerryan** [comment]
Thanks for the heads up @petersilva. I'll be watching the mailing list and the Sarracenia [repo](https://github.com/MetPX/sarracenia) for any upcoming changes. Is there an estimate on when v3 will be production ready?

**Comment by petersilva** [comment]
uh... there are two separate things... This issue is about v03 the message protocol.  That's in production in some deployments.  v03... the format is working and already heavily used (it is used within supercomputing for mirroring, which is the heaviest load we have.) It is also used in 
some experimental work with the international weather community.

For the public facing sites, it isn't a development thing... We just need to plan and proceed with others... it is a deployment thing... and not a high priority right now.. Need to get to it at some point, but we have a lot of other fish to fry.  When we get to it, we will likely do a transition that lasts months (with both v02 and v03 available at the same time.)

If you are referring to the v03_wip branch in the repo... that is about the next major version of Sarracenia python app, which is a very deep re-factor.  It is supposed to make things a lot better for everybody by making the code easier to work with and use. It will take a couple of months still.
Since you aren't using the python app, this doesn't affect you at all.

so... v02 will be around for some time... I would expect v03 to show up on public sites... perhaps in the fall?

**Comment by petersilva** [comment]
yeah... and fyi... the format is not going to change... it looks like it is staying the way it is.

**Comment by petersilva** [comment]
OK so this issue went through a lot over time.

* worked with the WMO for about two years... was looking for compatibility.
* in 2022, WMO turned out to only want web services, and specifically rejects file transfer, and specifically rejects things needed for core features.  So stopped working with them (different goals.)
* Continued work since then on HPC mirroring use case, the format evolved quite a bit... 
     * *integrity* became *identity* and format changed.
     * *fileOps* have been added.
     * *reports* have been added.
     * *retrievePath* has been added.
     * *blocks* field has been added for partitioned transfers.
     * geoJSON compatible *feature* and *geometry* is now available.
     * *content* allows for embedding file data within the message (discouraged, usually a bad idea.)

Current default format is here:

* https://metpx.github.io/sarracenia/Reference/sr_post.7.html

Now this format has evolved as the deployment of sr3 (the software) has been more and more deployed, hitting more and more use cases.  This is the default output format for sr3, and is used in the HPC mirroring use case internally, and is used when starting new flows that don´t interact with traditional ones.

The traditional datamarts are almost all using v02 message format, while migrating to sr3 (the software) as the means of publishing.  A migration to publishing v03 format messages is still expected to occur sometime.   Timing not clear.

**Comment by tannerryan** [comment]
Thanks for the quick update @petersilva. Will go through the Sarracenia v03 documentation and update for the new protocol. Happy New Year.
